### PR TITLE
include _quarto.yml if available

### PR DESCRIPTION
**Pull request by romainfrancois**
### Intent

addresses #10995 

### Approach

If a `_quarto.yml` file is found in the same directory as the `.qmd` file, then add it: 

```r
      if (file.exists(file.path(dirname(quartoSrcFile), "_quarto.yml"))) {
        file_list <- c(file_list, "_quarto.yml")
      }
```

### Automated Tests

added this: 

```r
test_that(".rs.rsconnectDeployList() includes _quarto.yml file (#10995 )", {
   dir.create(tf <- tempfile()); on.exit(unlink(tf, TRUE, TRUE))

   target <- file.path(tf, "test.qmd")
   writeLines("---\ntitle: quarto document\n---\n", target)
   writeLines('project:\n  title: "test"\n', file.path(tf, "_quarto.yml"))

   res <- .rs.rsconnectDeployList(target, FALSE, target)[[1L]]
   expect_true("_quarto.yml" %in% res)
})
```

### QA Notes

> Add additional information for QA on how to validate the change, paying special attention to the level of risk, adjacent areas that could be affected by the change, and any important contextual information not present in the linked issues. 

### Checklist

- [ ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [ ] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [x] This PR passes all local unit tests

### CODE REVIEW

**Comment by romainfrancois** [comment]
cc @toph-allen, should this rather be the `_quarto.yml` file in the project directory, instead of the `_quarto.yml` file in the same directory as the `.qmd` ?

**Comment by romainfrancois** [comment]
Alternatively, it could be a job for `rs_quartoFileResources()` ? cc @jjallaire 

```cpp
SEXP rs_quartoFileResources(SEXP targetSEXP)
{
   std::vector<std::string> resources;
   std::string target = r::sexp::safeAsString(targetSEXP);
   if (!target.empty())
   {
      FilePath qmdPath = module_context::resolveAliasedPath(target);
      json::Object jsonInspect;
      Error error = quartoInspect(
        string_utils::utf8ToSystem(qmdPath.getAbsolutePath()), &jsonInspect
      );
      if (!error)
      {
         jsonInspect["resources"].getArray().toVectorString(resources);
      }
      else
      {
         LOG_ERROR(error);
      }
   }

   r::sexp::Protect protect;
   return r::sexp::create(resources, &protect);
}
```

**Comment by jgutman** [comment]
> cc @toph-allen, should this rather be the `_quarto.yml` file in the project directory, instead of the `_quarto.yml` file in the same directory as the `.qmd` ?

Haven't seen an answer yet from @toph-allen, but I believe it should always be _quarto.yml in the root of the project directory, regardless of the exact location of the qmd file. I think even if your project has subfolders with quarto docs, you have just the one _quarto.yml with all the project level metadata at the root

**Review by jgutman — CHANGES_REQUESTED** [review]

**Comment by jgutman** [comment]
> @jgutman I don't have an authoritative answer to this, sorry. I'd check the quarto.org docs maybe, or find ask closer to the Quarto project?
> 
> The closest thing I can see to a section on _quarto.yml is this: https://quarto.org/docs/projects/quarto-projects.html#shared-metadata
> 
> Seems like it is at project level, rather than dir.

Thanks for the link! That seems to confirm my understanding that _quarto.yml is project level metadata and there should be only 1 per project. However, that link also reveals that there can be a directory level metadata file called _metadata.yml, which does live in the same subdirectly as the quarto doc to be published, and can contain additional metadata. 

@romainfrancois I think we should update this PR to 1) look for _quarto.yml in the project root, not the source file subdirectory, and 2) also look for a file called _metadata.yml in the source file subdirectory, and add that to the file list if it exists as well

**Comment by romainfrancois** [comment]
Thanks @jgutman. Modified now so that it relies on quarto to tell what the project is, with `quarto inspect` because IIUC this is orthogonal to rstudio projects. 

With this hierarchy: 

```
<rstudio_project>
  |--  _quarto.yml
  |--  main.qmd
  |-- sub
      |--  _metadata.yml
      |--  sub.qmd
```

I'm getting: 

image

image

The tests do stress more cases.

**Comment by romainfrancois** [comment]
One thing that might be a quarto issue is that, at least with files in the temporary directory, I'm getting these weird `project` results: 

With this code: 

```r
dir.create(tf <- tempfile()); 
dir.create(file.path(tf, "sub"))

main <- file.path(tf, "main.qmd")
writeLines("---\ntitle: quarto document\n---\n", main)

sub <- file.path(tf, "sub", "sub.qmd")
writeLines("---\ntitle: quarto document\n---\n", sub)

writeLines('project:\n  title: "test"\n', file.path(tf, "_quarto.yml"))

system(paste0("quarto inspect ", tf, "/main.qmd | grep project"))  
system(paste0("quarto inspect ", tf, "/sub/sub.qmd | grep project"))  
```

getting: 

```
> system(paste0("quarto inspect ", tf, "/main.qmd | grep project"))  
  "project": "../../../private/tmp/RtmpAMjixD/file1842c7157d908"
> system(paste0("quarto inspect ", tf, "/sub/sub.qmd | grep project"))  
  "project": "../../../../private/tmp/RtmpAMjixD/file1842c7157d908"
```

so I couldn't use `quartoInspect()` to implement `rs_quartoFileProject()`

**Comment by romainfrancois** [comment]
cc @kevinushey because you might have some views on `rs_quartoFileProject` re the json and the output `SEXP`.

**Comment by jjallaire** [comment]
If the `_quarto.yml` file is in a parent directory, is that properly handled during bundling/deployment? (i.e. how does it handle the relative path to something outside of the content directory)

**Review by jgutman — APPROVED** [review]
LGTM although I'm not super familiar with the `runQuarto({"inspect"...)` part of it

File: src/cpp/session/modules/SessionRSConnect.R [hunk 1]
```diff
@@ -384,6 +402,9 @@
    .Call("rs_quartoFileResources", target)
 })
 
+.rs.addFunction("quartoFileProject", function(target) {
+   .Call("rs_quartoFileProject", basename(target), dirname(target))
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionRSConnect.R:406]:
```suggestion
   .Call("rs_quartoFileProject", basename(target), dirname(target), PACKAGE = "(embedding)")
```

File: src/cpp/session/modules/quarto/SessionQuarto.cpp [hunk 2]
```diff
@@ -616,6 +616,40 @@ SEXP rs_quartoFileResources(SEXP targetSEXP)
    return r::sexp::create(resources, &protect);
 }
 
+SEXP rs_quartoFileProject(SEXP basenameSEXP, SEXP dirnameSEXP)
```
Comment (kevinushey) [inline src/cpp/session/modules/quarto/SessionQuarto.cpp:619]: I think this implementation needs to be guarded by `QUARTO_ENABLED` as some platforms don't support Quarto.

Comment (kevinushey) [reply src/cpp/session/modules/quarto/SessionQuarto.cpp:619]: Made those changes in https://github.com/rstudio/rstudio/commit/39de73fd61812d19e1f070188eea14ddc932d84d.

**Pull request merged**